mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td

```diff
@@ -635,7 +635,15 @@ def ThreadprivateOp : OpenMP_Op<"threadprivate"> {
     parallel region gets the value in encountering thread. They should be
     different in case the later one is optimized.
 
-    The `sym_addr` refers to the addressing of the symbol.
+    The private storage for threadprivate data is created in OpenMPIRBuilder
```

Comment:
That's because the pointer to the original global variable is also one argument of the runtime call `__kmpc_threadprivate_cached`.

```
LLVMIR-DAG: @_QFEi = internal global i32 undef
LLVMIR-DAG: @_QFEi.cache = common global i8** null
LLVMIR-DAG: [[CACHE0:%.*]] = call i8* @__kmpc_threadprivate_cached(%struct.ident_t* @[[GLOB0:[0-9]+]], i32 [[TMP0:%.*]], i8* bitcast (i32* @_QFEi to i8*), i64 4, i8*** @_QFEi.cache)
```

Reply:
Why can't the omp.threadprivate Operation model the threadprivate_cache call as it is. i.e the omp.threadprivate takes in the address of a global or similar variable and returns the cached threadprivate location?

Currently for the following Fortran code,
```
program test
  integer, save :: i
  integer :: x
  !$omp threadprivate(i)
  x = i

  !$omp parallel
    x = i
  !$omp end parallel
end
```
we have the following FIR + OpenMP
```
func @_QQmain() {
  %0 = fir.address_of(@_QFEi) : !fir.ref<i32>
  omp.threadprivate(%0 : !fir.ref<i32>) {sym_name = "_QFEi"}
  %1 = fir.alloca i32 {bindc_name = "x", uniq_name = "_QFEx"}
  %2 = fir.load %0 : !fir.ref<i32>
  fir.store %2 to %1 : !fir.ref<i32>
  omp.parallel  {
    %3 = fir.alloca i32 {bindc_name = "i", pinned, uniq_name = "_QFEi"}
    omp.threadprivate(%3 : !fir.ref<i32>) {sym_name = "_QFEi"}
    %4 = fir.load %3 : !fir.ref<i32>
    fir.store %4 to %1 : !fir.ref<i32>
    omp.terminator
  }
  return
}
fir.global internal @_QFEi : i32 {
  %0 = fir.undefined i32
  fir.has_value %0 : i32
}
```

I was asking whether we can have the following,
```
func @_QQmain() {
  %0 = fir.address_of(@_QFEi) : !fir.ref<i32>
  %tc1 = omp.threadprivate(%0 : !fir.ref<i32>) {sym_name = "_QFEi"}
  %1 = fir.alloca i32 {bindc_name = "x", uniq_name = "_QFEx"}
  %2 = fir.load %0 : !fir.ref<i32>
  fir.store %2 to %tc1 : !fir.ref<i32>
  omp.parallel  {
    %3 = fir.address_of(@_QFEi) : !fir.ref<i32>
    %tc2 = omp.threadprivate(%3 : !fir.ref<i32>) {sym_name = "_QFEi"}
    %4 = fir.load %3 : !fir.ref<i32>
    fir.store %4 to %tc2 : !fir.ref<i32>
    omp.terminator
  }
  return
}
fir.global internal @_QFEi : i32 {
  %0 = fir.undefined i32
  fir.has_value %0 : i32
}
```

Reply:
I think you are proposing this: (storing i to x instead of storing x to i in source code)
```
func @_QQmain() {
  %0 = fir.address_of(@_QFEi) : !fir.ref<i32>
  %tc1 = omp.threadprivate(%0 : !fir.ref<i32>) {sym_name = "_QFEi"}
  %1 = fir.alloca i32 {bindc_name = "x", uniq_name = "_QFEx"}
  %2 = fir.load %tc1 : !fir.ref<i32>
  fir.store %2 to %1 : !fir.ref<i32>
  omp.parallel  {
    %3 = fir.address_of(@_QFEi) : !fir.ref<i32>
    %tc2 = omp.threadprivate(%3 : !fir.ref<i32>) {sym_name = "_QFEi"}
    %4 = fir.load %tc2 : !fir.ref<i32>
    fir.store %4 to %1 : !fir.ref<i32>
    omp.terminator
  }
  return
}
```
This modelling looks better.

1. `{sym_name = "_QFEi"}` will not be need since `sym_name` was used to lookup the original GlobalOp. This modelling already has it.
2. I think `%3 = fir.address_of(@_QFEi) : !fir.ref<i32>` will be optimized. I have tried to create one new address_ofOp in parallel region long time ago. That's why I create pinned allocaOp in parallel region, and asking if we can add pinned attribute to address_ofOp. Without pinned attribute, this will become the following:
```
func @_QQmain() {
  %0 = fir.address_of(@_QFEi) : !fir.ref<i32>
  %tc1 = omp.threadprivate(%0 : !fir.ref<i32>) {sym_name = "_QFEi"}
  %1 = fir.alloca i32 {bindc_name = "x", uniq_name = "_QFEx"}
  %2 = fir.load %tc1 : !fir.ref<i32>
  fir.store %2 to %1 : !fir.ref<i32>
  omp.parallel  {
    %tc2 = omp.threadprivate(%0 : !fir.ref<i32>) {sym_name = "_QFEi"}
    %4 = fir.load %tc2 : !fir.ref<i32>
    fir.store %4 to %1 : !fir.ref<i32>
    omp.terminator
  }
  return
}
```
But based on your current proposed modelling, it seems we don't need pinned attribute since we will create `[[CACHE0:%.*]] = call i8* @__kmpc_threadprivate_cached(...)` and map this llvm Value to mlir Value of `%tc2`. So all in all, the final modelling should be like the following:
```
func @_QQmain() {
  %0 = fir.address_of(@_QFEi) : !fir.ref<i32>
  %tc1 = omp.threadprivate(%0 : !fir.ref<i32>)
  %1 = fir.alloca i32 {bindc_name = "x", uniq_name = "_QFEx"}
  %2 = fir.load %tc1 : !fir.ref<i32>
  fir.store %2 to %1 : !fir.ref<i32>
  omp.parallel  {
    %tc2 = omp.threadprivate(%0 : !fir.ref<i32>)
    %3 = fir.load %tc2 : !fir.ref<i32>
    fir.store %3 to %1 : !fir.ref<i32>
    omp.terminator
  }
  return
}
```

Reply:
Yes, what you are suggesting matches what I was trying to say.

Reply:
> Yes, what you are suggesting matches what I was trying to say.

Thanks for the confirmation. Will do it and update this PR later.